demos/tutorial/new_bc_2.cpp

```diff
@@ -0,0 +1,308 @@
+
+    CoordsRegion* clone() const override
+    {
+        return new CoordsRegion(m_func);
+    }
+
+    const samurai::LevelCellArray<dim> apply(const samurai::LevelCellArray<dim>& mesh) const override
+    {
+        samurai::LevelCellList<dim> lcl{mesh.level()};
+        auto set = samurai::difference(mesh, samurai::contraction(mesh));
+        samurai::for_each_cell(mesh, set, [&](auto& cell)
+        {
+            if (m_func(cell.center()))
+            {
+                lcl.add_cell(cell);
+            }
+            // samurai::static_nested_loop<dim, 0, 2>([&](auto stencil)
+            // {
+            //     if (m_func(cell.corner() + cell.length*stencil))
+            //     {
+            //         lcl.add_cell(cell);
+            //     }
+            // });
+        });
+        return lcl;
+    }
+
+    std::function<bool(const xt::xtensor_fixed<double, xt::xshape<dim>>&)> m_func;
+};
+
+template<std::size_t dim, class Set>
+struct SetRegion: public BcRegion<dim>
+{
+    SetRegion(const Set& set)
+    : m_set(set)
+    {}
+
+    SetRegion* clone() const override
+    {
+        return new SetRegion(m_set);
+    }
+
+    const samurai::LevelCellArray<dim> apply(const samurai::LevelCellArray<dim>& mesh) const override
+    {
+        return samurai::intersection(mesh, m_set);
+    }
+
+    Set m_set;
+};
+
+template<std::size_t dim, class F, class... CT>
+auto make_region(samurai::subset_operator<F, CT...> region)
+{
+    return std::make_unique<SetRegion<dim, samurai::subset_operator<F, CT...>>>(region);
+}
+
+template<std::size_t dim, class Func>
+auto make_region(Func&& func)
+{
+    return std::make_unique<CoordsRegion<dim>>(std::forward<Func>(func));
+}
+
+// Bc
+/////////////////////////////////////////////////////////////////////////////////////////
+template<std::size_t dim, class T, std::size_t size>
+struct Bc
+{
+    virtual ~Bc() = default;
+
+    template<class Bcvalue>
+    Bc(const Bcvalue& bcv)
+
+    {
+        p_bcvalue = std::make_unique<Bcvalue>(bcv);
+        p_bcregion = std::make_unique<Everywhere<dim>>();
+    }
+
+    Bc(const Bc& bc)
+    : p_bcvalue(bc.p_bcvalue->clone())
+    , p_bcregion(bc.p_bcregion->clone())
+    {
+    }
+
+    template<class Region>
+    auto on(const Region& region)
+    {
+        p_bcregion = make_region<dim>(region);
+        return *this;
+    }
+
+    auto get_lca(const samurai::LevelCellArray<dim>& mesh)
+    {
+        return p_bcregion->apply(mesh);
+    }
+
+    std::unique_ptr<BcValue<dim, T, size>> p_bcvalue;
+    std::unique_ptr<BcRegion<dim>> p_bcregion;
+};
+
+
+template<std::size_t dim, class T, std::size_t size>
+struct Dirichlet: public Bc<dim, T, size>
+{
+    using Bc<dim, T, size>::Bc;
+};
+
+template<std::size_t dim, class T, std::size_t size>
+struct Neumann: public Bc<dim, T, size>
+{
+    using Bc<dim, T, size>::Bc;
+};
+
+template<std::size_t dim, class T, std::size_t size>
+struct Robin: public Bc<dim, T, size>
+{
+    using Bc<dim, T, size>::Bc;
+};
```

Comment:
Why not add `apply()` inside those classes? Otherwise they will always be empty.
Plus, no need for dynamic_cast when `apply()` needs to be called.